### PR TITLE
kvs-watch: Support FLUX_KVS_WATCH_APPEND

### PR DESCRIPTION
**Pull request by chu11**
Support a new watch flag that will only respond with data appended to a key, not the full value of the key (#2054).

Notes:

- If an append does not occur, such as the key being overwritten by a new value, EINVAL is sent back to the user.  Better errno option?

- B/c it was easier, I elected to implement this watch via the key VALUES, not the valref treeobj array.  Because of this fact, appends can be faked.  e.g. the watch responses from

```
  flux kvs put key="a"
  flux kvs put --append key="b"
  flux kvs put --append key="c"
```

would be identical to watch results from

```
  flux kvs put key="a"
  flux kvs put key="ab"
  flux kvs put key="abc"
```

I felt this is the correct behavior to support.  Error in the watch would occur if a user instead did:

```
  flux kvs put key="a"
  flux kvs put key="b"
```

b/c it's clearly not an append.

I fixed up some stupid things along the way, including #2048.

### CODE REVIEW

**Comment by chu11** [comment]
good thing for coverage tests.  Missed some coverage, realized I had a missing test, and that missing test failed b/c I had a bug.  Re-pushed.

**Comment by garlick** [comment]
Nice that this isn't a lot of code!

I wonder if storing the full "value" (cumulative) in the watch state (in kvs-watch module) is needed?  If there were a lot of these going at once, that could eat up quite a lot of memory?

**Comment by chu11** [comment]
> I wonder if storing the full "value" (cumulative) in the watch state (in kvs-watch module) is needed? If there were a lot of these going at once, that could eat up quite a lot of memory?

Yeah, it's the one thing I was a bit iffy on.  Without it, the only check I would do is on data length.  So something like:

```
  flux kvs put key="a"
  flux kvs put key="bb"
```

would return a watch response of "a" and "b", since the length is ok.  But that seems really wrong.

We could of course do something clever like hash the value and see if the hash is identical?  So that guarantees us a max storage?  In fact, something clever like that could be used for `FLUX_KVS_WATCH_FULL` and `FLUX_KVS_WATCH_UNIQ` as well.

**Comment by garlick** [comment]
What about watching the treeobj, and just keeping an offset into the blobref array?   Stashing the most recently returned blobref might be sufficient to detect most non-append updates?  (And again, I don't think catching all possible cases of that is strictly necessary...)

**Comment by garlick** [comment]
>  What about watching the treeobj, and just keeping an offset into the blobref array

That was probably a bogus suggestion since you'd then have to go fetch the blobs, duplicating code in the kvs that can do that with the benefit of a local cache.

**Comment by garlick** [comment]
> Without it, the only check I would do is on data length

Might be OK.

Another idea (may not be a good one!): adding an offset to `lookup-plus`, to avoid fetching the same data over and over.

**Comment by chu11** [comment]
> That was probably a bogus suggestion since you'd then have to go fetch the blobs, duplicating code in the kvs that can do that with the benefit of a local cache.

Yeah, that was exactly the reason :-)

> Another idea (may not be a good one!): adding an offset to lookup-plus, to avoid fetching the same data over and over.

I had considered making a `lookup-plus-append` kind of special RPC to handle a lot of the internal work, but it seemed like more work than was necessary.  Given an previous treeobj for comparison, handle the lookup of new blobrefs to get the new value.

Another gotcha to recall.  Multiple appends can happen in the same kvs transaction, so a watch-append response could contain data from multiple blobrefs.  This is one of the gotchas that made me go with the way I implemented it. (which now that I mention it, I should have added a test for that, added that test and re-pushed).

> Without it, the only check I would do is on data length
> Might be OK.

So do we want to go with it?  I think the only thing is the language/doc of the feature would maybe have to be tweaked a bit.  Data past the previous data's length is returned.

**Comment by garlick** [comment]
> Multiple appends can happen in the same kvs transaction, so a watch-append response could contain data from multiple blobrefs. This is one of the gotchas that made me go with the way I implemented it. (which now that I mention it, I should have added a test for that, added that test and re-pushed).

I assumed this would be the case.  Are you saying you worked around that somehow?

My feeling is let's go without checking that already-returned content hasn't changed, unless you feel strongly about retaining that test?

**Comment by chu11** [comment]
> I assumed this would be the case. Are you saying you worked around that somehow?

No, just wanted to note that by comparing key values this is "automatic", but if we did it with treeobject valref indexes, there'd be some additional trickiness.

> My feeling is let's go without checking that already-returned content hasn't changed, unless you feel strongly about retaining that test?

Nah, I just think some of the documentation needs to be adjusted to note expectations.

**Comment by chu11** [comment]
rebased & re-pushed with the data comparison removed and a minor tweak to the manpage noting that only data length is looked at.

File: doc/man3/flux_kvs_lookup.adoc [hunk 1]
```diff
@@ -136,6 +136,14 @@ Specified along with FLUX_KVS_WATCH, this flag will alter watch
 behavior to only respond when _key_ is mentioned verbatim in a
 committed transaction and the value of the key has changed.
 
+FLUX_KVS_WATCH_APPEND::
+Specified along with FLUX_KVS_WATCH, this flag will alter watch
+behavior to only respond when _key_ is mentioned verbatim in a
+committed transaction and the key has been appended to.  The response
+will only contain the additional appended data.  Note that only data
+length is considered for appends and no guarantee is made that prior
+data hasn't have been overwritten.
```
Comment (garlick) [inline doc/man3/flux_kvs_lookup.adoc]: There's an extra "have" in this sentence.

File: src/modules/kvs-watch/kvs-watch.c [hunk 2]
```diff
@@ -267,6 +268,15 @@ static int handle_initial_response (flux_t *h,
         || (w->flags & FLUX_KVS_WATCH_UNIQ))
         w->prev = json_incref (val);
 
+    if (w->flags & FLUX_KVS_WATCH_APPEND) {
```
Comment (garlick) [inline src/modules/kvs-watch/kvs-watch.c]: Double parens are usually advised when testing a bit value inside an if.

File: src/modules/kvs-watch/kvs-watch.c [hunk 3]
```diff
@@ -311,6 +321,73 @@ static int handle_compare_response (flux_t *h,
...
+        if (flux_respond_pack (h, w->request, "{ s:O }", "val", new_val) < 0) {
+            json_decref (new_val);
+            flux_log_error (h, "%s: flux_respond_pack", __FUNCTION__);
+            return -1;
+        }
+
+        json_decref (new_val);
+    }
+
+    w->responded = true;
```
Comment (garlick) [inline src/modules/kvs-watch/kvs-watch.c]: set inside !w->responded conditional leg?

File: src/modules/kvs-watch/kvs-watch.c [hunk 4]
```diff
@@ -311,6 +321,73 @@ static int handle_compare_response (flux_t *h,
...
+        if (!(new_val = treeobj_create_val (new_data + w->append_offset,
+                                            new_offset - w->append_offset))) {
+            free (new_data);
+            return -1;
+        }
+
+        free (new_data);
+        w->append_offset = new_offset;
+
+        if (flux_respond_pack (h, w->request, "{ s:O }", "val", new_val) < 0) {
```
Comment (garlick) [inline src/modules/kvs-watch/kvs-watch.c]: use little 'o' (steals reference) and eliminate json_decref in success case?

File: src/modules/kvs-watch/kvs-watch.c (hunk 3 above)
Comment (chu11) [reply src/modules/kvs-watch/kvs-watch.c]: ahh, i copied this from the ```handle_compare_response()``` function, should move it in there too.

**Comment by chu11** [comment]
rebased & re-pushed with changes suggested by @garlick

**Comment by garlick** [comment]
Great!  Thanks!

**Pull request merged**